test/hummingbot/connector/derivative/bybit_perpetual/test_bybit_perpetual_user_stream_data_source.py

```diff
@@ -0,0 +1,350 @@
+import asyncio
+import json
+from unittest import TestCase
+from unittest.mock import patch, AsyncMock
+from typing import Optional
+
+from hummingbot.connector.derivative.bybit_perpetual.bybit_perpetual_user_stream_data_source import BybitPerpetualUserStreamDataSource
+from hummingbot.connector.derivative.bybit_perpetual.bybit_perpetual_auth import BybitPerpetualAuth
+import hummingbot.connector.derivative.bybit_perpetual.bybit_perpetual_constants as CONSTANTS
+from hummingbot.connector.derivative.bybit_perpetual.bybit_perpetual_websocket_adaptor import BybitPerpetualWebSocketAdaptor
+
+
+class BybitPerpetualUserStreamDataSourceTests(TestCase):
+    # the level is required to receive logs from the data source loger
+    level = 0
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.api_key = 'testAPIKey'
+        self.secret = 'testSecret'
+        self.log_records = []
+        self.ws_sent_messages = []
+        self.ws_incoming_messages = asyncio.Queue()
+        self.listening_task = None
+
+        self.data_source = BybitPerpetualUserStreamDataSource(auth_assistant=BybitPerpetualAuth(api_key=self.api_key, secret_key=self.secret))
+        self.data_source.logger().setLevel(1)
+        self.data_source.logger().addHandler(self)
+
+    def tearDown(self) -> None:
+        self.listening_task and self.listening_task.cancel()
+        if self.data_source._session is not None:
+            asyncio.get_event_loop().run_until_complete(self.data_source._session.close())
+        super().tearDown()
+
+    def handle(self, record):
+        self.log_records.append(record)
+
+    def _is_logged(self, log_level: str, message: str) -> bool:
+        return any(record.levelname == log_level and record.getMessage() == message
+                   for record in self.log_records)
+
+    async def _get_next_received_message(self, timeout: Optional[int] = None):
+        return await self.ws_incoming_messages.get()
+
+    def _create_ws_mock(self):
+        ws = AsyncMock()
+        ws.send_json.side_effect = lambda sent_message: self.ws_sent_messages.append(sent_message)
+        ws.receive_str.side_effect = self._get_next_received_message
+        ws.receive_json.side_effect = self._get_next_received_message
```

Comment:
receive_str should be assigned a string, and receive_json should be assigned a dictionary. They can't be both the same thing.